lumeer-core/src/main/java/io/lumeer/core/facade/ImportFacade.java

```diff
@@ -0,0 +1,187 @@
+import io.lumeer.api.model.Document;
+import io.lumeer.api.model.Permission;
+import io.lumeer.api.model.Permissions;
+import io.lumeer.api.model.Project;
+import io.lumeer.api.model.ResourceType;
+import io.lumeer.api.model.Role;
+import io.lumeer.core.model.SimplePermission;
+import io.lumeer.core.util.CodeGenerator;
+import io.lumeer.engine.api.data.DataDocument;
+import io.lumeer.storage.api.dao.CollectionDao;
+import io.lumeer.storage.api.dao.DataDao;
+import io.lumeer.storage.api.dao.DocumentDao;
+import io.lumeer.storage.api.exception.ResourceNotFoundException;
+
+import com.univocity.parsers.common.processor.RowListProcessor;
+import com.univocity.parsers.csv.CsvParser;
+import com.univocity.parsers.csv.CsvParserSettings;
+
+import java.io.StringReader;
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+
+@RequestScoped
+public class ImportFacade extends AbstractFacade {
+
+   public static final String FORMAT_CSV = "csv";
+
+   @Inject
+   private CollectionDao collectionDao;
+
+   @Inject
+   private DocumentDao documentDao;
+
+   @Inject
+   private DataDao dataDao;
+
+   public Collection importDocuments(final String format, final String data) {
+      List<Document> documents;
+
+      switch (format.toLowerCase()) {
+         case FORMAT_CSV:
+            documents = parseCSVFile(data);
+            break;
+         default:
+            documents = Collections.emptyList();
+      }
+
+      Collection collection = createImportCollection();
+      if (documents.isEmpty()) {
+         return collection;
+      }
+
+      documents.forEach(doc -> addDocumentMetadata(collection, doc));
+
+      List<Document> storedDocuments = documentDao.createDocuments(documents);
+
+      List<DataDocument> dataDocuments = new LinkedList<>();
+      for (int i = 0; i < documents.size(); i++) {
+         DataDocument dataDocument = documents.get(i).getData();
+         dataDocument.setId(storedDocuments.get(i).getId());
+         dataDocuments.add(dataDocument);
+      }
+      dataDao.createData(collection.getId(), dataDocuments);
+
+      return collection;
+   }
+
+   private void addDocumentMetadata(Collection collection, Document document) {
+      document.setCollectionId(collection.getId());
+      document.setCreatedBy(authenticatedUser.getCurrentUsername());
+      document.setCreationDate(LocalDateTime.now());
+      document.setDataVersion(DocumentFacade.INITIAL_VERSION);
+   }
+
+   private Collection createImportCollection() {
+      checkProjectWriteRole();
+      Set<String> names = collectionDao.getAllCollections().stream().map(Collection::getName).collect(Collectors.toSet());
+      Set<String> codes = collectionDao.getAllCollections().stream().map(Collection::getCode).collect(Collectors.toSet());
```

Comment:
Why do we make two database calls instead of just one? We should first load all collections and then map it to those sets.

Reply:
You can use `getAllCollectionCodes()` for this purpose. You do not really need collection names here since they are not unique. We can use the file name as a collection name for instance.